azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/MessagingFactory.java

```diff
@@ -422,24 +423,39 @@ public void onError(Exception error) {
 
     private class RunReactor implements Runnable {
         final private Reactor rctr;
+        final private ExecutorService executorService;
+
+        volatile boolean hasStarted;
 
-        public RunReactor(final Reactor reactor) {
+        public RunReactor(final Reactor reactor, final ExecutorService executorService) {
             this.rctr = reactor;
+            this.executorService = executorService;
+            this.hasStarted = false;
         }
 
         public void run() {
-            if (TRACE_LOGGER.isInfoEnabled()) {
+            if (TRACE_LOGGER.isInfoEnabled() && !this.hasStarted) {
                 TRACE_LOGGER.info("messagingFactory[%s], hostName[%s], info[%s]",
                         getClientId(),
                         getHostName(),
                         "starting reactor instance.");
             }
 
+            boolean yieldedReactor = false;
+
             try {
-                this.rctr.setTimeout(3141);
-                this.rctr.start();
-                while (!Thread.interrupted() && this.rctr.process()) {
+                if (!this.hasStarted) {
+                    this.rctr.setTimeout(3141);
+                    this.rctr.start();
+                    this.hasStarted = true;
                 }
+
+                if (!Thread.interrupted() && this.rctr.process()) {
+                    this.executorService.submit(this);
```

Comment:
>this.executorService.submit(this); [](start = 20, length = 34)

Handle the case where ExecutorService.shutdown() is initiated - to percolate the error all the up to the API Calls..